### PR TITLE
Add transform script and tests

### LINKED ISSUE #17: [Create] Transform/Cleaning Script for Incidents
## Task Description
- Develop a script for transforming and cleaning incidents data to prepare it for analysis and visualisation.
## User Stories
- As an engineer, I need a script to clean and transform raw incidents data into a structured format so that I can perform meaningful analysis.
- As an engineer, I need to create a script that can handle various data cleaning and transformation tasks efficiently.
## Relevant Files [If Available]
- Raw incidents data files

### PR DESCRIPTION
**Pull request by SaniyaShaikhh**
## Incident Transform Script 
I have added a transform script which contains several functions to process the data: a function to process the xml and extract the relevant information, a function to convert the timestamps to datetime objects, and a function to replace new line characters.

## Incident Transform Script Tests
I have created a test file "test_transform.py" which contains unit tests to test that each of the functions in the transform script return the correct objects. I have added a folder called test_data which contains an xml file and variables which are imported and used in the main test file.

## Other
- Updated the requirements.txt file to include modules I used to process the data
- Updated the extract.py file: imported the transform function and added it the on_message function such that each messaged is processed as it is received. 

closes  #17 #54

### CODE REVIEW

File: incidents/test_transform.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""Unit tests to test transform functions"""
+import pandas as pd
+
+from transform import process_xml, replace_new_line, convert_timestamps
+from test_data.test_vars import (OUTPUT_AFTER_PROCESSING_XML,
```
Comment (aribasyeda) [inline incidents/test_transform.py:5]: Excellant use of importing test varibales into your test file, makes the file easier to read and definately less clumpy

File: incidents/test_transform.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+"""Unit tests to test transform functions"""
```
Comment (aribasyeda) [inline incidents/test_transform.py:1]: Can add a space after the module docstring for styling purposes (and consistency with the other scripts)

File: incidents/test_transform.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                                 TIMESTAMP_OUTPUT_2)
+
+UK_TIMEZONE = 'Europe/London'
+
+with open("incidents/test_data/test_incident_data.xml", "r", encoding="utf-8") as file:
+    MSG = file.read()
+
+
+def test_process_xml_normal_input():
+    """tests process xml function if normal input string"""
```
Comment (aribasyeda) [inline incidents/test_transform.py:19]: Forgot to capitalise the 't'

File: incidents/test_transform.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    processed_msg = process_xml(MSG)
+    assert processed_msg == OUTPUT_AFTER_PROCESSING_XML
+    assert isinstance(processed_msg, list)
+
+
+def test_process_xml_bad_input():
```
Comment (aribasyeda) [inline incidents/test_transform.py:26]: here too

File: incidents/test_transform.py (hunk 3 above)
Comment (aribasyeda) [reply incidents/test_transform.py:19]: Docstring can be rephrased to add clarity

File: incidents/test_transform.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    """tests process xml function if empty input string"""
```
Comment (aribasyeda) [inline incidents/test_transform.py:27]: Similarly for this docstring - can be rephrased

File: incidents/test_transform.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    empty_msg = ''
+    assert process_xml(empty_msg) is None
+
+
+def test_replace_new_line():
+    """Tests replace new line character function"""
```
Comment (aribasyeda) [inline incidents/test_transform.py:34]: Can be more specific with what it tests e.g. It tests a normal/valid input

File: incidents/test_transform.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    processed_msg = process_xml(MSG)
+    data_frame = pd.DataFrame(processed_msg)
+    formatted_msg = replace_new_line(data_frame).to_dict(orient='list')
```
[the remark this answers is not in the document]
Comment (aribasyeda) [reply incidents/test_transform.py:38]: Using `.to_dict(orient='list')` is a clever approach! Alternative method to consider is keeping your data in the DataFrame format and directly comparing it to OUTPUT_AFTER_REPLACE using `pandas.testing.assert_frame_equal` but this method would need to format OUTPUT_AFTER_REPLACE as a DataFrame in your test setup

File: incidents/transform.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    for col in columns_to_convert:
+        if df[col].notnull().any():
+            df[col] = pd.to_datetime(df[col])
+            df[col] = df[col].dt.tz_convert(tz=timezone)
+
+    return df
+
+
+def transform_message(message):
```
Comment (aribasyeda) [inline incidents/transform.py:83]: Missing type hint

File: incidents/transform.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+"""Functions to transform and clean extracted data"""
```
Comment (aribasyeda) [inline incidents/transform.py:1]: Can add a space after the module docstring for styling purposes (and consistency with the other scripts)

File: incidents/extract.py [hunk 10]
```diff
@@ -26,6 +28,8 @@ def on_message(self, frame):
         """Executes when message is received"""
 
         logging.info('Received  message %s', frame.body)
+        cleaned_msg = transform_message(frame.body)
+        print(cleaned_msg)
```
[the remark this answers is not in the document]
Comment (aribasyeda) [reply incidents/extract.py:32]: Think you've forgotten the print message

**Comment by aribasyeda** [comment]
Solid code as always 💪 
What do you think about moving the `if __name__ == "__main__":` in extract.py to a separate pipeline/main.py file so that the main logic is ran in a script purely dedicated to running the pipeline. This will encourage SRP (i.e. separation of concerns) and make sure the code is focused

File: incidents/extract.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (SaniyaShaikhh) [reply incidents/extract.py:32]: @isaacschaessenscoleman Thanks Isaac! The way the connection works in the script is using environment variables: apologies, I should have made this clearer in my initial pull request. I will definitely update the documentation with what is required for the script to work. 

The variables are: 
- HOST - this is the messaging host for National Rail Knowledge Base live feed
- STOMP_PORT - the port to connect to
- USERNAME - credentials (from National Rail Data Portal account)
- PASSWORD - credentials (from National Rail Data Portal account)
- INCIDENTS_TOPIC - the messaging topic that needs to be subscribed to in order to receive the incident data

So when running the extract.py script the connection to the host (on the relevant port) is made first through the function `get_stomp_conn()`. The `connect_and_subscribe()` function then connects using the set of credentials and subscribes to the relevant topic (kb.incidents). Every time a message is received all of the code within the `on_message` function then gets executed. So as each message is received, the data gets transformed. The frame is just the message coming in from the live feed. I hope that clears things up a bit.

[the remark this answers is not in the document]
Comment (SaniyaShaikhh) [reply incidents/extract.py:32]: @aribasyeda I purposely left that in there for now so I can check that the messages are being transformed each time they are received. When I complete the next 2 scripts (loading into the database and the SNS alert system), I will need to edit extract.py to include these processes. I will remove all unnecessary print statements when everything is complete.

File: incidents/transform.py (hunk 8 above)
Comment (SaniyaShaikhh) [reply incidents/transform.py:83]: Thanks, I'll add that in!!

**Comment by SaniyaShaikhh** [comment]
> Solid code as always 💪 What do you think about moving the `if __name__ == "__main__":` in extract.py to a separate pipeline/main.py file so that the main logic is ran in a script purely dedicated to running the pipeline. This will encourage SRP (i.e. separation of concerns) and make sure the code is focused

That is a really good suggestion - I will implement this once I've sorted out the load and alert scripts.

**Review by aribasyeda — APPROVED** [review]
Nothing to add apart from great work!!!

**Pull request merged**